File: IDATT2101 Algoritmer og datastrukturer/cpp/task_4/2/Hashtable2.cpp
#include <chrono>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

using namespace std;
using namespace chrono;


class HashTable2{
    public:
        int *arr;
        int collisions;
        int length;

    bool checkPrime(int k){
        bool isPrime = true;
        for (int i = 2; i<k/2;i++){
            if((k%i) == 0){
                isPrime = false;
            }
        }
        return isPrime;
    }
    int getLength(int length, double times){
        int num =(int)(length*times);
        bool isPrime = false;
        while(!isPrime){
            isPrime = checkPrime(num);
            num++;
        }
        return isPrime?num-1: -1;
    }

    explicit HashTable2(int length){
        this->length = getLength(length,1.2);
        this-> arr = new int[this->length];
        this->collisions = 0;
        for (int i = 0; i < this->length; i++){
            arr[i] = 0;
        }
        
    }
    
    int hash(int key){
        return key%this->length;
    }
    int hash2(int key){
        return (key%(length-1)) + 1;
    }

    void insert(int key){
        int h = hash(key);
        if(arr[h] == 0){
            arr[h] = key;
        }else{
            int h2 = hash2(key);
            if(arr[h2] == 0){
                arr[h2] = key;
                return;
            }
            int pos = (h2)%this->length;
            while(true){
                if(arr[pos] == 0){
                    arr[pos] = key;
                    break;
                }
                collisions++;
                pos = (pos + h2)%this->length;
            }
        }
        
    }
    int get(int key){
        int h = hash(key);
        if(arr[h] == key){
            return arr[h];
        }else{
            int h2 = hash2(key);
            if(arr[h2] == key){
                return arr[h2]; 
            }
            int pos = (h2)%this->length;
            while(true){
                if(arr[pos] == 0){
                    return arr[pos];
                }
                collisions++;
                pos = (pos + h2)%this->length;
            }
        }
    }
};

int main() {
    int length = 10000000;
    int find = 212121;
    HashTable2 ht(length);
    ht.insert(find);
    double time = 0;
    int *nrs = new int[length];
    int nr;
     timespec start, stop;
    for(int i = 0; i<(length/2)-1;i++ ){
            nr = (int)(rand()%length*10);
            nrs[i] = nr;
            nrs[(length/2)-1-i] = nr;
    }
    for(int i = 0; i<length-1;i++ ){
        clock_gettime(CLOCK_THREAD_CPUTIME_ID, &start);
        ht.insert(nrs[i]);
        clock_gettime(CLOCK_THREAD_CPUTIME_ID, &stop);
        time += stop.tv_nsec - start.tv_nsec ;
    }
        printf("size: %d \n", length);
        printf("time: %f ms \n", time/1000000);
        printf("a = %f \n",(double)length/ht.length);
        printf("collisions: %d \n", ht.collisions);
        printf("%d\n",ht.get(find));
}

File: IDATT2101 Algoritmer og datastrukturer/java/task_4/2/HashTable2.java
import java.util.HashMap;

public class HashTable2 {
    private int arr[];
    private int collisions;
    final int PRIME = 7 ;
    double A = (Math.sqrt(5.0)-1.0)/2;

    boolean checkPrime(int k){
        boolean isPrime = true;
        for (int i = 2; i<k/2;i++){
            if((k%i) == 0){
                isPrime = false;
            }
        }
        return isPrime;
    }
    int getLength(int length, double times){
        int num =(int)(length*times);
        boolean isPrime =false;
        while(!isPrime){
            isPrime = checkPrime(num);
            num++;
        }
        return isPrime?num-1: -1;
    }
    public HashTable2(int length){
        //this.arr = new int[(int)Math.pow(2,Math.ceil(Math.log(length)/Math.log(2)))];
        this.arr = new int[getLength(length,1.2)];
        this.collisions = 0;
    }

    public int multiHash(int t){
        double nr = t*A;
        nr -= (int)nr;
        return (int)(Math.abs(nr)*arr.length);
    }
    public int divHash(int t){
        return t%(arr.length);
    }
    public int h_2(int t){
        return ((2*Math.abs(t))+1)%(arr.length);
    }
    public  int h_2_p(int t){
        return PRIME-(t%PRIME);//(t%(PRIME-1))+1;//PRIME-(t%PRIME);
    }
    public int probe(int index, int index2, int i){
       return (index + index2*i)%(arr.length);
    }

    public void insert(int t){
        int h = divHash(t);
        if(arr[h] == 0){
            arr[h] = t;
        }else{
            int h2 = h_2_p(t);
            int n = probe(h,h2,1);
            while(true){
                n = (n+h2)%arr.length;
                if(arr[n]==0){
                    arr[n] = t;
                    break;
                }
                collisions++;              
            }
        }
    }
    public int get(int t){
        int h = divHash(t);
        if(arr[h] == t){
            return arr[h];
        }else{
            int h2 = h_2_p(t);
            int n = probe(h,h2,1);
            while(true){
                n = (n+h2)%arr.length;
                if(arr[n]==t){
                    return arr[n];
                }
            }
        }
    }

    public static void main(String[] args) {
        int length = 10000000;
        int find = 212121;
        HashTable2 ht = new HashTable2(length);
        ht.insert(find);
        long start, end;
        long totalT = 0;
        int nrs[] = new int[length];
        int nr;
        for(int i = 0; i<(length/2)-1;i++ ){
            nr = (int)(Math.random()*length*10);
            nrs[i] = nr;
            nrs[(length/2)-1-i] = nr;
        }
        for(int i = 0; i<length-1;i++ ){
            start = System.nanoTime();
            ht.insert(nrs[i]);
            end = System.nanoTime();
            totalT += end-start;
        }
        System.out.println("size: " + length);
        System.out.println("time: " + totalT/1000000 +"ms");
        System.out.println("a =" + (double)length/ht.arr.length);
        System.out.println("collisions: "+ ht.collisions);
        System.out.println(ht.get(find));
        totalT = 0;
        HashMap<Integer,Integer> h = new HashMap<>();
        for(int i = 0; i<length-1;i++ ){
            start = System.nanoTime();
            h.put(nrs[i],nrs[i]);
            end = System.nanoTime();
            totalT += end-start;
        }
        System.out.println("Java Map time: " + totalT/1000000 +"ms");

    }

}

